### PR TITLE
Fix/28759 billing shipping addresses

### LINKED ISSUE #28759: Billing County is copied into Shipping Address
**Prerequisites (mark completed items with an [x]):**
- [x ] I have have carried out troubleshooting steps and I believe I have found a bug.
- [ x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
If your billing address has a county but shipping address doesn't then that field gets copied over and merged into the shipping address.

I have tested this on two live sites with very different setups running the latest woo (4.8.0).
Also tested on 4.0.1 which does not have this behaviour.
And tested on a totally fresh demo site (4.8.0) at http://woocommerce.oxwebdev.com/ - feel free to test here

**Expected behavior**
Billing fields should never be merged into a shipping address.

**Actual behavior**
A single field (County) is merged from billing to shipping, creating an invalid shipping address

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
1. Create a new account
2. Add a billing address including a county
3. Add a shipping address without a county
4. Attempt to buy an item
5. Merge is visible on the cart screen and also in the inputs that are revealed when toggling "Ship to a different address"

**Screenshots**

screen-2021-01-08_12-40-04

screen-2021-01-08_12-40-24

screen-2021-01-08_12-40-46

**Isolating the problem (mark completed items with an [x]):**
- [x ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x ] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
We use the [WooCommerce System Status Report](https://docs.woocommerce.com/document/understanding-the-woocommerce-system-status-report/) to help us evaluate the issue. 
Without this report we won't be able to fully evaluate this issue.
<details>
`
### WordPress Environment ###

WordPress address (URL): http://woocommerce.oxwebdev.com/wp
Site address (URL): http://woocommerce.oxwebdev.com
WC Version: 4.8.0
REST API Version: ✔ 4.8.0
WC Blocks Version: ✔ 3.8.1
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 1.7.3
Log Directory Writable: ❌ To allow logging
make /var/www/woocommerce.oxwebdev.com/woocommerce/web/app/uploads/wc-logs/ writable or define a custom WC_LOG_DIR.

WP Version: 5.6
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.29 (Ubuntu)
PHP Version: 7.2.24-0ubuntu0.18.04.6
PHP Post Max Size: 64 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.58.0
OpenSSL/1.1.1

SUHOSIN Installed: –
MySQL Version: 5.7.31-0ubuntu0.18.04.1
Max Upload Size: 64 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ❌ Your server does not support the mbstring functions - this is required for better character encoding. Some fallbacks will be used instead for it.
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 4.8.0
WC Database Prefix: wp_
Total Database Size: 4.68MB
Database Data Size: 2.88MB
Database Index Size: 1.80MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.52MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wfblockediplog: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfblocks7: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wfconfig: Data: 0.09MB + Index: 0.00MB + Engine InnoDB
wp_wfcrawlers: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wffilechanges: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wffilemods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfhits: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wfhoover: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wfissues: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wfknownfilelist: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wflivetraffichuman: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wflocs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wflogins: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wfls_2fa_secrets: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wfls_settings: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfnotifications: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfpendingissues: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wfreversecache: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfsnipcache: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wfstatus: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wftrafficrates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

customize_changeset: 1
page: 10
post: 2
product: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (2) ###

WooCommerce: by Automattic – 4.8.0
Wordfence Security: by Wordfence – 7.4.14

### Must Use Plugins (3) ###

Bedrock Autoloader: by Roots – 1.0.3
Disallow Indexing: by Roots – 1.0.0
Register Theme Directory: by Roots – 1.0.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #10 - /index.php/shop/
Cart: #11 - /index.php/cart/
Checkout: #12 - /index.php/checkout/
My account: #13 - /index.php/my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 3.1.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 5
Oldest: 2021-01-07 14:48:37 +0000
Newest: 2021-01-08 12:40:35 +0000

`
</details>

### PR DESCRIPTION
**Pull request by barryhughes**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

In some countries the state (county) field is optional—the UK is one such example. That means we might have a saved billing address _with_ a county, and a saved shipping address _without_ (or similar). 

In these cases we should not copy the billing state/county to the shipping state/county: that action is really only appropriate if we don't already have a shipping address. Note that this problem also applies to the country field (though, it is easiest to see in relation to the state/county field).

Closes #28759.

### How to test the changes in this Pull Request:

1. Start by confirming you can see the original problem (checkout [current trunk code](https://github.com/woocommerce/woocommerce/commit/855f48d53b9a83c109891340b2b71bf9772db89f), or else the `5.1.0` tag).
    1. Create a customer account and ensure shipping is set up.
    2. Add and save billing and shipping address, supplying a county for the billing address but not for the shipping address.
    3. Now purchase a product (it's easiest to see if you buy a simple physical product).
    4. Note that the shipping address appearing on the cart page includes the billing county.
2. Now checkout this branch.
    1. Start by clearing/deleting the user session† and logging back in, otherwise the shipping address will be inherited from the first test (complete with inappropriately added state/county). 
    2. Then, simply repeat the test. The billing county is not copied over to the shipping address.

† _If you have your own snippet to kill the user session and dump the cart contents that's probably fine, or you could try using [this snippet](https://gist.github.com/barryhughes/d17134dc83dd216555115182879da5f4) (updating the user ID, and if necessary the table prefix, accordingly) then flushing the cache._

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - If we have a non-empty shipping address then do not overwrite the state or country fields with billing address data.

### CODE REVIEW

**Comment by roykho** [comment]
In my testing it is still being copied when I go to checkout.


And here is my addresses for billing and shipping

**Comment by barryhughes** [comment]
Hey @roykho ... apologies, I missed a key step in my testing instructions (now updated). Can you try again?

**Comment by roykho** [comment]
I used incognito and thought that would be a cleared session but maybe not. Ok let me test again. And BTW you can use this tool to clear the session

**Comment by barryhughes** [comment]
I'm not sure incog by itself is enough, if you then log back into the same account, because of the cart persistence that happens via stored user meta data. Also, cool—was totally unaware of that tool. TIL 🙃

File: includes/class-wc-customer.php [hunk 1]
```diff
@@ -242,6 +242,25 @@ public function has_calculated_shipping() {
 		return $this->get_calculated_shipping();
 	}
 
+	/**
+	 * Indicates if the customer has a non-empty shipping address.
+	 *
+	 * Note that this does not indicate if the customer's shipping address
+	 * is complete, only that one or more fields are populated.
+	 *
+	 * @return bool
```
Comment (roykho) [inline includes/class-wc-customer.php:251]: Maybe it would be good to add `@since 5.3.0` versioning.

Comment (barryhughes) [reply includes/class-wc-customer.php:253]: Added.

**Review by roykho — APPROVED** [review]
LGTM! Tested well and thanks for writing the tests!

**Pull request merged**